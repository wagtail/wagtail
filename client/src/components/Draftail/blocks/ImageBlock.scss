@mixin wagtail-label-overrides {
    float: none;
    width: initial;
    padding: 0;
}

.ImageBlock__field {
    @include wagtail-label-overrides;
    display: block;
    margin-bottom: $button-spacing * 2;
    color: $color-white;
    cursor: pointer;
}

@mixin wagtail-input-overrides {
    border: 0;
    font-size: inherit;
}

.ImageBlock__field__input {
    @include wagtail-input-overrides;
    font-size: 1rem;
    background-color: $color-white;
    color: $color-text-base;

    &[readonly] {
        color: transparentize($color-text-base, 0.5);
    }
}

.ImageBlock__alt {
    @include font-smoothing;
    font-size: 0.875rem;
}
