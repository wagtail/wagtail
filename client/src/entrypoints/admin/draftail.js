import * as Draftail from 'draftail';
import draftail, {
  Link,
  Document,
  ImageBlock,
  EmbedBlock,
} from '../../components/Draftail/index';

/**
 * Entry point loaded when the Draftail editor is in use.
 */

// Expose Draftail package as a global.
window.Draftail = Draftail;
// Expose module as a global.
window.draftail = draftail;

// Plugins for the built-in entities.
const plugins = [
  {
    type: 'DOCUMENT',
    source: draftail.DocumentModalWorkflowSource,
    decorator: Document,
  },
  {
    type: 'LINK',
    source: draftail.LinkModalWorkflowSource,
    decorator: Link,
  },
  {
    type: 'IMAGE',
    source: draftail.ImageModalWorkflowSource,
    block: ImageBlock,
  },
  {
    type: 'EMBED',
    source: draftail.EmbedModalWorkflowSource,
    block: EmbedBlock,
  },
];

plugins.forEach(draftail.registerPlugin);
