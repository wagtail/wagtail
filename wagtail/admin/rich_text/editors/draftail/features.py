from django.forms import Media

# Feature objects: these are mapped to feature identifiers within the rich text
# feature registry (wagtail.core.rich_text.features). Each one implements
# a `construct_options` method which modifies an options dict as appropriate to
# enable that feature.

# Additionally, a Feature object defines a media property
# (https://docs.djangoproject.com/en/stable/topics/forms/media/) to specify css/js
# files to import when the feature is active.


class Feature:
    def __init__(self, js=None, css=None):
        self.js = js
        self.css = css

    @property
    def media(self):
        return Media(js=self.js, css=self.css)


class BooleanFeature(Feature):
    """
    A feature which is enabled by a boolean flag at the top level of
    the options dict
    """
    def __init__(self, option_name, **kwargs):
        super().__init__(**kwargs)
        self.option_name = option_name

    def construct_options(self, options):
        options[self.option_name] = True


class ListFeature(Feature):
    """
    Abstract class for features that are defined in a list within the options dict.
    Subclasses must define option_name
    """
    def __init__(self, data, **kwargs):
        super().__init__(**kwargs)
        self.data = data

    def construct_options(self, options):
        if self.option_name not in options:
            options[self.option_name] = []

        options[self.option_name].append(self.data)


class EntityFeature(ListFeature):
    """A feature which is listed in the entityTypes list of the options"""
    option_name = 'entityTypes'


class BlockFeature(ListFeature):
    """A feature which is listed in the blockTypes list of the options"""
    option_name = 'blockTypes'


class InlineStyleFeature(ListFeature):
    """A feature which is listed in the inlineStyles list of the options"""
    option_name = 'inlineStyles'
