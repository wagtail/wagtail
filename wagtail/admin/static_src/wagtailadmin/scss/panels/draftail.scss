@import 'wagtailadmin/scss/helpers';

@import '../../../../../../client/src/components/Draftail/Draftail';
