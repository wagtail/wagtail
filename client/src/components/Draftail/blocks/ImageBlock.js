import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { DraftUtils } from 'draftail';

import { STRINGS } from '../../../config/wagtailConfig';

import MediaBlock from '../blocks/MediaBlock';

/**
 * Editor block to preview and edit images.
 */
class ImageBlock extends Component {
  constructor(props) {
    super(props);

    this.changeAlt = this.changeAlt.bind(this);
  }

  changeAlt(e) {
    const { block, blockProps } = this.props;
    const { editorState, onChange } = blockProps;

    const data = {
      alt: e.target.value,
    };

    onChange(DraftUtils.updateBlockEntity(editorState, block, data));
  }

  render() {
    const { blockProps } = this.props;
    const { entity, onRemoveEntity } = blockProps;
    const { src, alt } = entity.getData();
    const altLabel = `${STRINGS.ALT_TEXT}: “${alt || ''}”`;

    return (
      <MediaBlock {...this.props} src={src} alt="">
        <p className="ImageBlock__alt">{altLabel}</p>

        <button className="button button-secondary no Tooltip__button" onClick={onRemoveEntity}>
          {STRINGS.DELETE}
        </button>
      </MediaBlock>
    );
  }
}

ImageBlock.propTypes = {
  block: PropTypes.object.isRequired,
  blockProps: PropTypes.shape({
    editorState: PropTypes.object.isRequired,
    entity: PropTypes.object,
    onChange: PropTypes.func.isRequired,
  }).isRequired,
};

export default ImageBlock;
