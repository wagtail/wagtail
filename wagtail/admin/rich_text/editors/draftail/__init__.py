import json

from django.forms import Media, widgets

from wagtail.admin.edit_handlers import RichTextFieldPanel
from wagtail.admin.rich_text.converters.contentstate import ContentstateConverter
from wagtail.admin.staticfiles import versioned_static
from wagtail.core.rich_text import features as feature_registry


class DraftailRichTextArea(widgets.HiddenInput):
    template_name = 'wagtailadmin/widgets/draftail_rich_text_area.html'
    is_hidden = False

    # this class's constructor accepts a 'features' kwarg
    accepts_features = True

    def get_panel(self):
        return RichTextFieldPanel

    def __init__(self, *args, **kwargs):
        # note: this constructor will receive an 'options' kwarg taken from the WAGTAILADMIN_RICH_TEXT_EDITORS setting,
        # but we don't currently recognise any options from there (other than 'features', which is passed here as a separate kwarg)
        kwargs.pop('options', None)
        self.options = {}

        self._media = Media(js=[
            versioned_static('wagtailadmin/js/draftail.js'),
        ], css={
            'all': [versioned_static('wagtailadmin/css/panels/draftail.css')]
        })

        self.features = kwargs.pop('features', None)
        if self.features is None:
            self.features = feature_registry.get_default_features()

        for feature in self.features:
            plugin = feature_registry.get_editor_plugin('draftail', feature)
            if plugin:
                plugin.construct_options(self.options)
                self._media += plugin.media

        self.converter = ContentstateConverter(self.features)

        default_attrs = {'data-draftail-input': True}
        attrs = kwargs.get('attrs')
        if attrs:
            default_attrs.update(attrs)
        kwargs['attrs'] = default_attrs

        super().__init__(*args, **kwargs)

    def format_value(self, value):
        # Convert database rich text representation to the format required by
        # the input field
        value = super().format_value(value)

        if value is None:
            value = ''

        return self.converter.from_database_format(value)

    def get_context(self, name, value, attrs):
        context = super().get_context(name, value, attrs)
        context['widget']['options_json'] = json.dumps(self.options)
        return context

    def value_from_datadict(self, data, files, name):
        original_value = super().value_from_datadict(data, files, name)
        if original_value is None:
            return None
        return self.converter.to_database_format(original_value)

    @property
    def media(self):
        return self._media
