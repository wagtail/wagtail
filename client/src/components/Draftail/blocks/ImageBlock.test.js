import React from 'react';
import { shallow } from 'enzyme';

import { DraftUtils } from 'draftail';

import ImageBlock from '../blocks/ImageBlock';

describe('ImageBlock', () => {
  it('renders', () => {
    expect(
      shallow(
        <ImageBlock
          block={{}}
          blockProps={{
            editorState: {},
            entityType: {},
            entity: {
              getData: () => ({
                src: 'example.png',
              }),
            },
            onChange: () => {},
          }}
        />
      )
    ).toMatchSnapshot();
  });

  it('no data', () => {
    expect(
      shallow(
        <ImageBlock
          block={{}}
          blockProps={{
            editorState: {},
            entityType: {},
            entity: {
              getData: () => ({}),
            },
            onChange: () => {},
          }}
        />
      )
    ).toMatchSnapshot();
  });

  it('alt', () => {
    expect(
      shallow(
        <ImageBlock
          block={{}}
          blockProps={{
            editorState: {},
            entityType: {},
            entity: {
              getData: () => ({
                src: 'example.png',
                alt: 'Test',
              }),
            },
            onChange: () => {},
          }}
        />
      )
    ).toMatchSnapshot();
  });

  it('changeAlt', () => {
    jest.spyOn(DraftUtils, 'updateBlockEntity');
    DraftUtils.updateBlockEntity.mockImplementation(e => e);

    const onChange = jest.fn();
    const wrapper = shallow(
      <ImageBlock
        block={{}}
        blockProps={{
          editorState: {},
          entityType: {},
          entity: {
            getData: () => ({
              src: 'example.png',
              alt: 'Test',
            }),
          },
          onChange,
        }}
      />
    );

    // // Alt field is readonly for now.
    wrapper.instance().changeAlt({
      target: {
        value: 'new alt',
      }
    });
    // wrapper.find('[type="text"]').simulate('change', {
    //   target: {
    //     value: 'new alt',
    //   },
    // });

    expect(onChange).toHaveBeenCalled();
    expect(DraftUtils.updateBlockEntity).toHaveBeenCalledWith(
      expect.any(Object),
      {},
      expect.objectContaining({ alt: 'new alt' })
    );

    DraftUtils.updateBlockEntity.mockRestore();
  });
});
