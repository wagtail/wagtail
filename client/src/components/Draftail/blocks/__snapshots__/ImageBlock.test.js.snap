// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageBlock alt 1`] = `
<MediaBlock
  alt=""
  block={Object {}}
  blockProps={
    Object {
      "editorState": Object {},
      "entity": Object {
        "getData": [Function],
      },
      "entityType": Object {},
      "onChange": [Function],
    }
  }
  src="example.png"
>
  <p
    className="ImageBlock__alt"
  >
    Alt text: “Test”
  </p>
  <button
    className="button button-secondary no Tooltip__button"
  >
    Delete
  </button>
</MediaBlock>
`;

exports[`ImageBlock no data 1`] = `
<MediaBlock
  alt=""
  block={Object {}}
  blockProps={
    Object {
      "editorState": Object {},
      "entity": Object {
        "getData": [Function],
      },
      "entityType": Object {},
      "onChange": [Function],
    }
  }
  src={null}
>
  <p
    className="ImageBlock__alt"
  >
    Alt text: “”
  </p>
  <button
    className="button button-secondary no Tooltip__button"
  >
    Delete
  </button>
</MediaBlock>
`;

exports[`ImageBlock renders 1`] = `
<MediaBlock
  alt=""
  block={Object {}}
  blockProps={
    Object {
      "editorState": Object {},
      "entity": Object {
        "getData": [Function],
      },
      "entityType": Object {},
      "onChange": [Function],
    }
  }
  src="example.png"
>
  <p
    className="ImageBlock__alt"
  >
    Alt text: “”
  </p>
  <button
    className="button button-secondary no Tooltip__button"
  >
    Delete
  </button>
</MediaBlock>
`;
